File: codeWars/katas/TTT55_find_a_matrix_in_a_matrix/TTT55_find_a_matrix_in_a_matrix.ts
// T.T.T.55: Find a matrix in a matrix
export {};

function findMatrixIn(matrix: number[][]): number[][] {
  const rows: number = matrix.length;
  const cols: number = matrix[0].length;

  let maxSubMatrixSize: number = 0;
  let subMatrixRow: number = -1;
  let subMatrixCol: number = -1;

  for (let i = 0; i < rows; i++) {
    for (let j = 0; j < cols; j++) {
      let countZeros: number = 0;
      let countOnes: number = 0;

      for (let x = i; x < rows; x++) {
        for (let y = j; y < cols; y++) {
          if (matrix[x][y] === 0) {
            countZeros++;
          } else {
            countOnes++;
          }

          if (
            countZeros === countOnes &&
            (x - i + 1) * (y - j + 1) > maxSubMatrixSize
          ) {
            maxSubMatrixSize = (x - i + 1) * (y - j + 1);
            subMatrixRow = i;
            subMatrixCol = j;
          }
        }
      }
    }
  }

  if (maxSubMatrixSize === 0) {
    return [];
  }

  const subMatrix: number[][] = [];

  for (
    let i = subMatrixRow;
    i < subMatrixRow + Math.sqrt(maxSubMatrixSize);
    i++
  ) {
    const row: number[] = [];

    for (
      let j = subMatrixCol;
      j < subMatrixCol + Math.sqrt(maxSubMatrixSize);
      j++
    ) {
      row.push(matrix[i][j]);
    }

    subMatrix.push(row);
  }

  return subMatrix;
}

const matrix1: number[][] = [
  [1, 1, 1],
  [0, 0, 0],
  [1, 1, 1],
];
const result1: number[][] = findMatrixIn(matrix1);
console.log(result1);
// Output: [
//   [1, 1],
//   [0, 0]
// ]

const matrix2: number[][] = [
  [1, 1, 1, 0],
  [0, 0, 0, 1],
  [1, 1, 1, 0],
  [0, 0, 0, 1],
];
const result2: number[][] = findMatrixIn(matrix2);
console.log(result2);
// Output: [
//   [1, 1, 1, 0],
//   [0, 0, 0, 1],
//   [1, 1, 1, 0],
//   [0, 0, 0, 1]
// ]

File: codeWars/katas/TTT55_find_a_matrix_in_a_matrix/TTT55_find_a_matrix_in_a_matrix.py
# T.T.T.55: Find a matrix in a matrix


def findMatrixIn(matrix):
    rows = len(matrix)
    cols = len(matrix[0])

    maxSubMatrixSize = 0
    subMatrixRow = -1
    subMatrixCol = -1

    for i in range(rows):
        for j in range(cols):
            countZeros = 0
            countOnes = 0

            for x in range(i, rows):
                for y in range(j, cols):
                    if matrix[x][y] == 0:
                        countZeros += 1
                    else:
                        countOnes += 1

                    if (
                        countZeros == countOnes
                        and (x - i + 1) * (y - j + 1) > maxSubMatrixSize
                    ):
                        maxSubMatrixSize = (x - i + 1) * (y - j + 1)
                        subMatrixRow = i
                        subMatrixCol = j

    if maxSubMatrixSize == 0:
        return []

    subMatrix = []

    for i in range(subMatrixRow, subMatrixRow + int(maxSubMatrixSize**0.5)):
        row = []

        for j in range(subMatrixCol, subMatrixCol + int(maxSubMatrixSize**0.5)):
            row.append(matrix[i][j])

        subMatrix.append(row)

    return subMatrix


matrix1 = [[1, 1, 1], [0, 0, 0], [1, 1, 1]]
result1 = findMatrixIn(matrix1)
print(result1)
# Output: [
#   [1, 1],
#   [0, 0]
# ]

matrix2 = [[1, 1, 1, 0], [0, 0, 0, 1], [1, 1, 1, 0], [0, 0, 0, 1]]
result2 = findMatrixIn(matrix2)
print(result2)
# Output: [
#   [1, 1, 1, 0],
#   [0, 0, 0, 1],
#   [1, 1, 1, 0],
#   [0, 0, 0, 1]
# ]
